eventstream/interface.go

```diff
@@ -46,11 +46,18 @@ var ErrDone = errors.New("no more items in iterator")
 // Iterator iterates an event stream.  To be used concurrently with Publish operations.
 //
 // Unlike Promises, Iterators are stateful and should not be shared across go routines.
-type Iterator interface {
+type Iterator[T any] interface {
 	// Next returns the next event in the stream.
 	// - Returns (<event>, nil) when the next event is published.
-	// - Returns (nil, ErrDone) when the stream is exhausted.
-	// - Returns (nil, ctx.Err()) if the context is cancelled.
+	// - Returns (zero, ErrDone) when the stream is exhausted.
+	// - Returns (zero, ctx.Err()) if the context is cancelled.
 	// Blocks until one of these three outcomes occurs.
-	Next(ctx context.Context) (interface{}, error)
+	Next(ctx context.Context) (T, error)
+
+	// Consume iterates the remained of the stream, calling the provided callback with each successive value.
+	// - Returns (nil) when the stream is exhausted.
+	// - Returns (ctx.Err()) if the context is cancelled.
+	// - Returns (<error>) if the callback returns any non-nil error.
+	// Blocks until one of these three outcomes occurs.
+	Consume(ctx context.Context, callback func(context.Context, T) error) error
```

Comment:
Should we provide a way to signal a non-erroring stopping condition? Something like `io.EOF` or even an eventstream package specific error.

Reply:
I see your point.  I'm a little torn tho.

On the one hand, it would be easy enough make it so that if the callback returns `eventstream.ErrDone`, it gets translated until a null.  But it's also more complicated to explain in doc.

On the other hand, since the caller's error is transparently handed back to them, it's easy enough for the caller to just use a sentinel error themselves.

Reply:
👋 

IMO the existing behavior ("returns nil when the stream is exhausted") seems pretty reasonable

Reply:
👋 😁

Reply:
Would this wording be clear enough?
```golang
// Returns (nil) when the stream is exhausted or the callback returns `eventstream.ErrDone`
```

Reply:
fair enough